src/ripple/rpc/handlers/Tx.cpp

```diff
@@ -97,11 +96,68 @@ Json::Value doTx (RPC::Context& context)
     if (!isHexTxID (txid))
         return rpcError (rpcNOT_IMPL);
 
-    auto txn = context.app.getMasterTransaction ().fetch (
-        from_hex_text<uint256>(txid), true);
+    ClosedInterval<uint32_t> range;
+
+    auto rangeProvided = context.params.isMember (jss::min_ledger) &&
+        context.params.isMember (jss::max_ledger);
+
+    if (rangeProvided)
+    {
+        try
+        {
+            auto const& min = context.params[jss::min_ledger].asUInt ();
+            auto const& max = context.params[jss::max_ledger].asUInt ();
+
+            constexpr uint16_t MAX_RANGE = 1000;
+
+            if (max < min)
+                return rpcError (rpcINVALID_LGR_RANGE);
+
+            if (max - min > MAX_RANGE)
+                return rpcError (rpcEXCESSIVE_LGR_RANGE);
+
+            range = ClosedInterval<uint32_t> (min, max);
+        }
+        catch (...)
+        {
+            // One of the calls to `asUInt ()` failed.
+            return rpcError (rpcINVALID_LGR_RANGE);
+        }
+    }
+
+    using pointer = Transaction::pointer;
+
+    auto searchedAll = false;
+    auto ec {rpcSUCCESS};
+    pointer txn;
+
+    if (rangeProvided)
+    {
+        boost::variant<Transaction::pointer, bool> v =
+            context.app.getMasterTransaction().fetch(
+                from_hex_text<uint256>(txid), range, ec);
+
+        if (v.which () != 0 || !boost::get<pointer> (v))
+            searchedAll = boost::get<bool> (v);
```

Comment:
Its actually not possible for both of these conditions to be true:

-  `v.which() != 0`
-  `boost::get<pointer> (v) == nullptr` 

But I will change the logic as you've indicated to simplify the code.

Reply:
I meant `v.which() == 0` and the pointer part of the variant is a nullpointer then line 141 throws 
(I suspect you meant that in your replay as well). If that can't happen (it looks like it only happens on a serialization error), then we shouldn't check for it. If we do check for it, we need to do something sensible when we detect it.

Reply:
Yeah, that also can’t happen. Basically in this branch of the code we’ll never have a ‘nullptr’ because we’d be return a ‘bool’ in the ‘variant’ instead. But yeah, I’ll clean up the logic here to be cleaner.